File: src/main/java/arc/arc/autobuild/gui/BuildingGui.java
package arc.arc.autobuild.gui;

import arc.arc.ARC;
import arc.arc.autobuild.BuildingManager;
import arc.arc.autobuild.ConstructionSite;
import arc.arc.configs.Config;
import arc.arc.configs.ConfigManager;
import arc.arc.util.GuiUtils;
import arc.arc.util.TextUtil;
import com.github.stefvanschie.inventoryframework.adventuresupport.TextHolder;
import com.github.stefvanschie.inventoryframework.gui.GuiItem;
import com.github.stefvanschie.inventoryframework.gui.type.ChestGui;
import com.github.stefvanschie.inventoryframework.pane.OutlinePane;
import com.github.stefvanschie.inventoryframework.pane.Pane;
import com.github.stefvanschie.inventoryframework.pane.StaticPane;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Material;
import org.bukkit.entity.HumanEntity;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.util.List;

public class BuildingGui extends ChestGui {
    Player player;
    ConstructionSite site;
    boolean youSure = false;
    BukkitTask renameTask;
    Config config = ConfigManager.of(ARC.plugin.getDataPath(), "auto-build.yml");

    public BuildingGui(Player player, ConstructionSite site) {
        super(3, TextHolder.deserialize("&8Потверждение строительства"), ARC.plugin);
        this.player = player;
        this.site = site;

        setupBg();
        setupButtons();
    }

    private void setupButtons() {
        StaticPane staticPane = new StaticPane(0, 1, 9, 1);

        ItemStack confirmStack = new ItemStack(config.material("building-gui.confirm-material", Material.PAPER));
        ItemStack cancelStack = new ItemStack(config.material("building-gui.cancel-material", Material.RED_STAINED_GLASS_PANE));

        ItemMeta confirmMeta = confirmStack.getItemMeta();
        confirmMeta.displayName(TextUtil.strip(Component.text("Прогресс строительства", NamedTextColor.GOLD)));
        double progress = site.getProgress();
        int percentage = (int) (progress * 100);
        confirmMeta.lore(List.of(TextUtil.strip(Component.text("> " + percentage + "%", NamedTextColor.GRAY))));
        new BukkitRunnable() {
            @Override
            public void run() {
                if (site.getState() != ConstructionSite.State.BUILDING) {
                    this.cancel();
                    getViewers().forEach(HumanEntity::closeInventory);
                    return;
                }
                double progress = site.getProgress();
                int percentage = (int) (progress * 100);
                ItemMeta meta = confirmStack.getItemMeta();
                meta.lore(List.of(TextUtil.strip(Component.text("> " + percentage + "%", NamedTextColor.GRAY))));
                confirmStack.setItemMeta(meta);
                update();
            }
        }.runTaskTimer(ARC.plugin, 20L, 60L);
        confirmStack.setItemMeta(confirmMeta);

        ItemMeta cancelMeta = cancelStack.getItemMeta();
        cancelMeta.displayName(config.componentDef("building-gui.cancel-name", "<red>Отменить постройку"));
        int cancelModelData = config.integer("building-gui.cancel-model-data", 0);
        if (cancelModelData != 0) cancelMeta.setCustomModelData(cancelModelData);
        cancelStack.setItemMeta(cancelMeta);

        GuiItem confirmItem = new GuiItem(confirmStack, inventoryClickEvent -> inventoryClickEvent.setCancelled(true));
        staticPane.addItem(confirmItem, 2, 0);

        GuiItem cancelItem = new GuiItem(cancelStack, inventoryClickEvent -> {
            inventoryClickEvent.setCancelled(true);

            if (youSure) {
                if (renameTask != null && !renameTask.isCancelled()) renameTask.cancel();
                BuildingManager.cancelConstruction(site);
                inventoryClickEvent.getWhoClicked().closeInventory();
            } else {
                youSure = true;
                ItemMeta meta = confirmStack.getItemMeta();
                meta.displayName(config.componentDef("building-gui.cancel-name", "<red>Отменить постройку"));
                meta.lore(config.componentListDef("building-gui.cancel-lore", List.of(
                        "<gray>Вы уверены, что хотите отменить строительство?",
                        "<gray>Вы не вернете книгу"
                )));
                cancelStack.setItemMeta(meta);

                renameTask = new BukkitRunnable() {
                    @Override
                    public void run() {
                        ItemMeta cancelMeta = cancelStack.getItemMeta();
                        cancelMeta.displayName(config.componentDef("building-gui.cancel-name", "<red>Отменить постройку"));
                        cancelStack.setItemMeta(cancelMeta);
                        youSure = false;
                        update();
                    }
                }.runTaskLater(ARC.plugin, 5 * 20L);

                this.update();
            }

        });
        staticPane.addItem(cancelItem, 6, 0);


        Material fastFinishMaterial = config.material("building-gui.fast-finish-material", Material.BLAZE_POWDER);
        GuiItem fastFinish = new GuiItem(new ItemStack(fastFinishMaterial), inventoryClickEvent -> {
            inventoryClickEvent.setCancelled(true);
            site.finishBuildState();
            inventoryClickEvent.getWhoClicked().closeInventory();
        });
        if (player.hasPermission("arc.build.fast")) {
            staticPane.addItem(fastFinish, 4, 0);
        }
        this.addPane(staticPane);
    }

    private void setupBg() {
        OutlinePane pane = new OutlinePane(0, 0, 9, 3);
        pane.addItem(GuiUtils.background());
        pane.setRepeat(true);
        pane.setPriority(Pane.Priority.LOWEST);
        this.addPane(pane);
    }
}

File: src/main/kotlin/ru/arc/autobuild/gui/BuildingGui.kt
package ru.arc.autobuild.gui

import com.github.stefvanschie.inventoryframework.adventuresupport.TextHolder
import com.github.stefvanschie.inventoryframework.gui.GuiItem
import com.github.stefvanschie.inventoryframework.gui.type.ChestGui
import com.github.stefvanschie.inventoryframework.pane.OutlinePane
import com.github.stefvanschie.inventoryframework.pane.Pane
import com.github.stefvanschie.inventoryframework.pane.StaticPane
import com.github.stefvanschie.inventoryframework.pane.util.Slot
import org.bukkit.Material
import org.bukkit.entity.HumanEntity
import org.bukkit.entity.Player
import ru.arc.ARC
import ru.arc.autobuild.BuildConfig
import ru.arc.autobuild.BuildingManager
import ru.arc.autobuild.ConstructionSite
import ru.arc.autobuild.ConstructionState
import ru.arc.core.ScheduledTask
import ru.arc.core.delayed
import ru.arc.core.repeating
import ru.arc.core.ticks
import ru.arc.util.GuiUtils
import ru.arc.util.TextUtil
import ru.arc.util.fromConfig
import ru.arc.util.guiItem

/**
 * GUI shown during active construction.
 * Shows progress and allows cancellation.
 */
class BuildingGui(
    private val player: Player,
    private val site: ConstructionSite
) : ChestGui(
        3,
        TextHolder.deserialize(TextUtil.toLegacy(BuildConfig.BuildingGui.title)),
        ARC.instance,
    ) {
    private var youSure = false
    private var renameTask: ScheduledTask? = null
    private var progressTask: ScheduledTask? = null

    init {
        setupBackground()
        setupButtons()

        setOnClose {
            progressTask?.cancel()
            renameTask?.cancel()
            progressTask = null
            renameTask = null
        }
    }

    private fun setupBackground() {
        val pane = OutlinePane(9, 3, Pane.Priority.LOWEST).apply {
            addItem(GuiUtils.background())
            setRepeat(true)
        }
        addPane(Slot.fromXY(0, 0), pane)
    }

    private fun setupButtons() {
        val pane = StaticPane(9, 1)
        val buildConfig = BuildConfig.config()

        // Progress indicator
        val percentage = (site.progress * 100).toInt()
        val progressItem =
            guiItem(Material.PAPER) {
                fromConfig(buildConfig, "building-gui.progress")
            }
        BuildingGuiPresentation.progress(buildConfig, percentage).applyTo(progressItem.item)

        // Auto-update progress using Task DSL
        progressTask =
            repeating(period = 60.ticks, delay = 20.ticks) {
                if (site.state != ConstructionState.Building) {
                    cancel()
                    viewers.forEach(HumanEntity::closeInventory)
                    return@repeating
                }
                val newPercentage = (site.progress * 100).toInt()
                BuildingGuiPresentation.progress(buildConfig, newPercentage).applyTo(progressItem.item)
                update()
            }

        pane.addItem(progressItem, 2, 0)

        // Cancel button with confirmation
        val normalCancel = BuildingGuiPresentation.item(buildConfig, "building-gui.cancel")
        val confirmCancel = BuildingGuiPresentation.item(buildConfig, "building-gui.cancel-confirm")
        lateinit var cancelItem: GuiItem
        cancelItem =
            guiItem(Material.RED_STAINED_GLASS_PANE) {
                fromConfig(buildConfig, "building-gui.cancel")
                onClick { event ->
                    event.isCancelled = true
                    if (youSure) {
                        renameTask?.takeIf { !it.isCancelled }?.cancel()
                        BuildingManager.cancelConstruction(site)
                        event.whoClicked.closeInventory()
                    } else {
                        youSure = true
                        confirmCancel.applyTo(cancelItem.item)

                        renameTask =
                            delayed((5 * 20).ticks) {
                                youSure = false
                                normalCancel.applyTo(cancelItem.item)
                                update()
                            }

                        update()
                    }
                }
            }
        normalCancel.applyTo(cancelItem.item)

        pane.addItem(cancelItem, 6, 0)

        // Fast finish (admin only)
        if (player.hasPermission("arc.build.fast")) {
            pane.addItem(
                guiItem(Material.BLAZE_POWDER) {
                    fromConfig(buildConfig, "building-gui.fast-finish")
                    onClick { event ->
                        site.finishInstantly()
                        event.whoClicked.closeInventory()
                    }
                },
                4,
                0,
            )
        }

        addPane(Slot.fromXY(0, 1), pane)
    }
}
